### PR TITLE
[installer]: allow use of external container registry

### LINKED ISSUE #6565: Support GCP image registry

### PR DESCRIPTION
**Pull request by MrSimonEmms**
## Description

Configure external container registry

## Related Issue(s)

Fixes #6565 

## How to test

Get create container registry in GCP and get a service account key

To create secret:
```
kubectl create secret docker-registry \
  registry-secret \
  --docker-server=gcr.io/<projectId>/gitpod \
  --docker-username=_json_key \
  --docker-password="$(cat /path/to/credentials.json)"
```

Then change the config to:
```
containerRegistry:
  inCluster: false
  external:
    url: gcr.io/<projectId>/gitpod
    certificate:
      kind: secret
      name: registry-secret
```

Deploy the changes and create a workspace

## Release Notes

```release-note
Allow use of external container registry
```

### CODE REVIEW

**Comment by roboquat** [comment]
@MrSimonEmms: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **mrsimonemms** after the PR has been reviewed.

Associated issue: *#6565*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: installer/pkg/components/proxy/configmap.go
Comment (MrSimonEmms) [inline installer/pkg/components/proxy/configmap.go:118]: FYI, this section has been rearranged to make the registry proxy conditional

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **csweichel** after the PR has been reviewed.

Associated issue: *#6565*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: installer/pkg/components/blobserve/deployment.go
Comment (csweichel) [inline installer/pkg/components/blobserve/deployment.go:25]:
```suggestion
	} else if ctx.Config.ContainerRegistry.External != nil {
```

File: installer/pkg/components/image-builder-mk3/configmap.go
Comment (csweichel) [inline installer/pkg/components/image-builder-mk3/configmap.go:29]:
```suggestion
	} else if ctx.Config.ContainerRegistry.External != nil {
```

} else { return fmt.Error }

File: installer/pkg/components/image-builder-mk3/configmap.go
Comment (csweichel) [inline installer/pkg/components/image-builder-mk3/configmap.go:30]:
```suggestion
		registryName = strings.TrimSuffix(ctx.Config.ContainerRegistry.External.URL, "/")
```

File: installer/pkg/components/image-builder-mk3/deployment.go
Comment (csweichel) [inline installer/pkg/components/image-builder-mk3/deployment.go:40]:
```suggestion
	} else if ctx.Config.ContainerRegistry.External != nil {
```

File: installer/pkg/components/registry-facade/daemonset.go
Comment (csweichel) [inline installer/pkg/components/registry-facade/daemonset.go:66]:
```suggestion
	} else if ctx.Config.ContainerRegistry.External != nil {
```

**Review by csweichel — COMMENTED** [review]
Found a few minor things that would need fixing. 

Is there an issue for introducing the secret validation checks?

**Comment by MrSimonEmms** [comment]
> Is there an issue for introducing the secret validation checks?

Yes - https://github.com/gitpod-io/gitpod/issues/6627

File: installer/pkg/components/blobserve/deployment.go
Comment (MrSimonEmms) [reply installer/pkg/components/blobserve/deployment.go:25]: Ok. I did this on the basis that it was either/or (protected by validation), but not against defensive coding

**Review by csweichel — APPROVED** [review]
LGTM

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: f0d88d09fc558ec8a4bfce3ed5358fe36e23f1d8</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6621#pullrequestreview-803552043" title="Approved">csweichel</a>*

Associated issue: *#6565*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
